### PR TITLE
python3 porting based on v1.1.8

### PR DESCRIPTION
**Pull request by madeddy**
- works since ~April 2021 on apps with RenPy engine **up to** version 7.4.11; later also as base for v8.0.x
- only issues are known and unrelated to the port like the "at transform" bug or warnings about "unknown user-statements" which are logical
- changes on the _readme_ should be ignored/discarded (they're just for info about the branches)
- `un.rpyc` submodule does not work (**no** changes on it)

### CODE REVIEW

**Comment by madeddy** [comment]
I explained in both of my posts that the `un.rpyc` submodule does not work.

**Comment by zmook** [comment]
I would sincerely love to get my hands on these changes, since *most* games fail decompilation now with the existing master branch. I've tried cloning madeddy/unrpyc, but I cannot quite figure out how to merge the py3_port@v1.1.8 and the Renpy_8_7.5_compat_fix@v1.1.8 branches without getting conflicts. And I think we need both, yes? If CensoredUsername is not around to process these pull requests, could you either provide a merged branch in  the madeddy fork, or notes on what I need to do myself?

**Comment by madeddy** [comment]
> I would sincerely love to get my hands on these changes...
>... but I cannot quite figure out how to merge the py3_port@v1.1.8 and the Renpy_8_7.5_compat_fix@v1.1.8 branches without getting conflicts. And I think we need both, yes? 

You're correct, you need both parts for RenPy v8.0.x based apps. However for v7.5.x you need just the compat fix, because there is RenPy still on Python 2! Not sure about the new RenPy 7.6.x(Py2) and v8.1.x(Py3), there could be additional changes needed.(We will have to wait and see...)

As for your request, i did already complete branches of this code for RenPy 8/Py3: https://github.com/madeddy/unrpyc/tree/py3_v1.1.8 and the v1.2.0 also but has some extra (non necessary luxus) changes.

You should look at the readme's of the branches. Good luck! Greets

**Comment by eirnym** [comment]
I should say, that most pre-compiled renpy 7.x games works well on RenPy 8.2.x nighlty without any recompilation, Some Python modules obviously won't be working as they have Python2 bytecode. There's not many of them though

**Comment by eirnym** [comment]
@madeddy The decompiller works well (1.2.0_dev at least) 

I suggest following fixes (could be done in separate smaller PRs)

1. For Python3 (with or without if) to use `dict`instead of `OrderedDict` as `dict` guarantees the order since Python 3.7
      * it's useful to minimise CPU and memory overhead of the `OrderedDict` if not needed. 
3. Don't use processes if single file supplied.
      * it's often some resources has not been freed, and there's no point of the process pool anyway.
      * I'd just add an `if` statement just after file list
5. Don't write a file if decompilation/de-ast has been failed by default.
      * For a user it's better to have an original `rpyc` if processing failed and repeat this process if needed.
      * To achieve this it's possible to pass `io.BytesIO` instead of file and write when no errors or requested
      * Interactive mode and/or additional parameters could be an option
7. Add an option to skip decompilation if a decompiled file exists. Normal mode, without a warning.
      * proposed argument names: `no-clobber` or `-C`
8. Pass arguments by name if  too many of them
      * example: call of `decompile_rpyc` in `worker` function
9. Some new `renpy.ast` objects are missing, like `renpy.ast.Camera`. There's not many new objects and they are not so popular, but some `rpyc`files can't be decompiled because of this.

File: decompiler/util.py [hunk 1]
```diff
@@ -433,18 +433,18 @@ def __init__(self, needs_space, reorderable=False):
         self.reorderable = reorderable
 
     def append(self, *args):
-        self.words.extend(filter(None, args))
+        self.words.extend([_f for _f in args if _f])
```
Comment (Gouvernathor) [inline decompiler/util.py:436]: Why this change ? filter is good, and I think it works on this instance

**Review by Gouvernathor — APPROVED** [review]

**Review by eirnym — APPROVED** [review]

File: decompiler/util.py (hunk 1 above)
Comment (eirnym) [reply decompiler/util.py:436]: I'd argue with your statement. Filer is more clear method and for those who is unfamiliar with it, I'd recommend to read `help(filter)` at least, for further reading Python tutorial and library reference. 

Additionally, filter is optimised to be fast

Comment (madeddy) [reply decompiler/util.py:436]: If it's faster then its a good argument and we have two votes against the generator. I'm not a hardass and sit on it just because. Even if i like the gen way. Going to revert. Thanks folks

File: decompiler/astdump.py [hunk 2]
```diff
@@ -274,10 +273,10 @@ def ind(self, diff_indent=0, ast=None):
         # shouldn't indent in case there's only one or zero objects in this object to print
         if ast is None or len(ast) > 1:
             self.indent += diff_indent
-            self.p(u'\n' + self.indentation * self.indent)
+            self.p('\n' + self.indentation * self.indent)
 
     def p(self, string):
         # write the string to the stream
-        string = unicode(string)
+        string = str(string)
```
Comment (madeddy) [inline decompiler/astdump.py:280]: Do we in py3 need this line?

Comment (madeddy) [reply decompiler/astdump.py:283]: Did some tests debug print in there and it was as thought always <class 'str'>, so i think i will test a while without the line before i commit it.

**Comment by madeddy** [comment]
> I suggest following fixes (could be done in separate smaller PRs)

Sry i missed last year somehow this post with the suggestions. Will look into it. Could you please move this into a separate issue? As this pull here is just the py3 port and your suggestions going more for extension/expansion. Thanks.


> I suggest looking at https://github.com/Gouvernathor/unrpyc/tree/dev3 

Yes, thank you. I discovered it already last year and used some of the code if that's ok with you.

**Comment by eirnym** [comment]
@madeddy yes, this would make sense after PR is merged

**Comment by eirnym** [comment]
Why PR has been closed?

**Comment by madeddy** [comment]
> Why PR has been closed?

Good question. As i was it myself i can only deduct is was some backlash as i set my fork back and the connected branch disappeared accidentally.

I reopen it but its moot as the pull was to old to be integrated.

**Comment by madeddy** [comment]
@eirnym Since the version for py3 is almost ready, you can perhaps offer your list again in [GH diss](https://github.com/CensoredUsername/unrpyc/discussions) if you want.
NOTE: 6) is moot and 4) do we have already in unrpyc ... i think 🤔

**Comment by CensoredUsername** [comment]
We're up to date now, so this pull request is no longer necessary.
